### PR TITLE
MIP22: fixes

### PR DESCRIPTION
**Pull request by xmxanuel**
**Content**
- fix events to custom events
- rename ONE to RAY
- changed tranche approval `uint(-1)`
- rm pool approval
- rename owner to operator
- adjusted unit tests

-  filing functions for vow, end and join
-  return funds to operator after end has settled
-  add revert messages
-  operator can no longer call tell

### CODE REVIEW

File: src/mgr.sol [hunk 1]
```diff
@@ -127,20 +150,20 @@ contract TinlakeManager is LibNote {
         ilk = ilk_;
         wards[msg.sender] = 1;
```
Comment (MrChico) [inline src/mgr.sol:158]:
```suggestion
emit Rely(msg.sender); 
```

Comment (xmxanuel) [reply src/mgr.sol:158]: I adopted it from the MIP21
https://github.com/livnev/rwa-example/pull/8/files#diff-20efe103664fbcfe47bafc2b8deb1cc4fcb9939f45fe32d91c2b0dee0ddc551aR31

Comment (MrChico) [reply src/mgr.sol:158]: see https://github.com/centrifuge/tinlake-maker-lib/pull/18#discussion_r565448088

File: src/mgr.sol [hunk 2]
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
 // not only DROP as an ERC20 balance in this contract, but also what's currently
 // undergoing redemption from the Tinlake pool.
 
-contract TinlakeManager is LibNote {
+contract TinlakeManager {
     // --- Auth ---
-    mapping (address => uint256) public wards;
-    function rely(address usr) external note auth { require(live, "TinlakeMgr/not-live"); wards[usr] = 1; }
-    function deny(address usr) external note auth { require(live, "TinlakeMgr/not-live"); wards[usr] = 0; }
+     mapping (address => uint256) public wards;
```
Comment (gbalabasquer) [inline src/mgr.sol]: nit: extra spacing

File: src/mgr.sol [hunk 3]
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
...
-        require(msg.sender == owner, "TinlakeMgr/owner-only");
+    modifier operatorOnly {
+        require(msg.sender == operator, "TinlakeMgr/operator-only");
         _;
     }
 
-    // The owner manages the cdp, but is not authorized to call kick or cage.
-    address public owner;
+    // Events
+    event Rely(address usr);
```
Comment (gbalabasquer) [inline src/mgr.sol]: I think we could set all addresses as `indexed` in the different events.

File: src/mgr.sol [hunk 4]
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
...
+    event Wipe(uint wad);
+    event Join(uint wad);
+    event Exit(uint wad);
+    event NewOperator(address usr);
+    event Tell(uint wad);
+    event Unwind(uint payBack);
+    event Sink(uint ink, uint tab);
+    event Recover(uint recovered, uint payBack);
+    event Cage();
+    event File(bytes32 what, address data);
```
Comment (gbalabasquer) [inline src/mgr.sol]: `what` might be `indexed` as well

File: src/mgr.sol [hunk 5]
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
...
+        require(msg.sender == operator, "TinlakeMgr/operator-only");
         _;
     }
 
-    // The owner manages the cdp, but is not authorized to call kick or cage.
-    address public owner;
+    // Events
+    event Rely(address usr);
+    event Deny(address usr);
+    event Draw(uint wad);
```
Comment (gbalabasquer) [inline src/mgr.sol]: nit: if possible make all the `uint` to `uint256`.

Comment (gbalabasquer) [reply src/mgr.sol]: Same for `int256`. I think some of them that we had changed, have been overwritten in this PR.

File: src/mgr.sol [hunk 6]
```diff
@@ -159,64 +186,80 @@ contract TinlakeManager is LibNote {
...
         vat.frob(ilk, address(this), address(this), address(this), 0, -int256(dart));
+        emit Wipe(wad);
     }
 
     // --- Administration ---
-    function setOwner(address newOwner) external ownerOnly note {
-        owner = newOwner;
+    function setOperator(address newOperator) external operatorOnly  {
+        operator = newOperator;
+        emit NewOperator(newOperator);
```
Comment (gbalabasquer) [inline src/mgr.sol]: Hmm maybe better calling it as `SetOperator`

File: src/mgr.sol [hunk 7]
```diff
@@ -159,64 +186,80 @@ contract TinlakeManager is LibNote {
...
         gem.approve(dst, uint256(-1));
         live = false;
+        emit Migrate(dst);
+    }
+
+    function file(bytes32 what, address data) public auth {
+        require(live, "TinlakeManager/not-live");
+        emit File(what, data);
+        if (what == "vow") vow = data;
+        else if (what == "join") daiJoin = JoinLike(data);
```
Comment (gbalabasquer) [inline src/mgr.sol]: I'd suggest call the param also `daiJoin`.

File: src/mgr.sol [hunk 8]
```diff
@@ -110,38 +133,42 @@ contract TinlakeManager is LibNote {
...
         safe = true;
         live = true;
         glad = true;
 
         dai.approve(daiJoin_, uint256(-1));
         vat.hope(daiJoin_);
-        gem.approve(pool_, uint256(-1));
-        gem.approve(tranche, uint256(-1));
+
+        tranche = tranche_;
```
Comment (gbalabasquer) [inline src/mgr.sol:168]: I know this comes from our comments, so actually directing this comment to @godsflaw and @brianmcmichael. Is it really worth it to save the variable to do partial approvals on `tell`? I mean if we can actually verify that `tranche` can only take `gem` when `tell` is called by governance we could avoid having to store this one.

File: src/mgr.sol [hunk 9]
```diff
@@ -159,64 +186,80 @@ contract TinlakeManager is LibNote {
...
+        else if (what == "join") daiJoin = JoinLike(data);
+        else if (what == "end")  end = EndLike(data);
+        else revert("Vat/file-unrecognized-param");
     }
 
     // --- Liquidation ---
-    function tell() public note {
-        require(safe);
-        require(wards[msg.sender] == 1 || (msg.sender == owner && !live), "TinlakeManager/not-authorized");
+    function tell() public auth {
```
Comment (gbalabasquer) [inline src/mgr.sol:255]: It seems `tell` can't be called anymore by the operator if shutdown happens. Is that an intentional change?
If this change was actually wrong, couldn't we even make that this is called by anyone (if shutdown happens)?

File: src/mgr.sol (hunk 8 above)
Comment (xmxanuel) [reply src/mgr.sol:168]: The tranche `redeemOrder` can only be triggered by Maker governance.  This is the normal way for DROP token holder to redeem their tokens. 
The mgr calls the `operator.redeemOrder`
redeemOrder in operator
https://github.com/centrifuge/tinlake/blob/develop/src/lender/operator.sol#L67
redeemOrder in tranche
https://github.com/centrifuge/tinlake/blob/develop/src/lender/tranche.sol#L128

I think we should just keep the `amount` approval then there is definitely no way that Tinlake
can access it.

In Tinlake we use a contract called `clerk` (operator in mgr terminology)  to interact with the `mgr`.
For all methods we require the following to be true:
```javascript
 function mkrActive() public view returns (bool) {
        return mgr.safe() && mgr.glad() && mgr.live();
    }
```
See: https://github.com/centrifuge/tinlake/blob/develop/src/lender/adapters/mkr/clerk.sol#L117

File: src/mgr.sol (hunk 9 above)
Comment (xmxanuel) [reply src/mgr.sol:255]: Yes, previously the only way to set the live flag to false was through the `migrate` call and by calling `cage` but cage requires a tell call beforehand. 

In the end the operator doesn't to have the power to trigger the `tell`.

That was the reason why @MrChico removed it.

@MrChico , correct?

Comment (gbalabasquer) [reply src/mgr.sol:255]: Hmm not really, `cage` can be called by anyone if shutdown happens in the MCD system. Prior to this PR, the operator could run `tell` if the `!live`.

File: src/mgr.sol (hunk 3 above)
Comment (gbalabasquer) [reply src/mgr.sol]: I've seen `indexed` has been added for several events but still missing for `Rely` and `Deny`

File: src/mgr.sol (hunk 9 above)
Comment (MrChico) [reply src/mgr.sol:255]: Yes, removing the ability for the operator to call `tell` was intentional. The operator can de-risk the position in other ways, and removing this privilege for the operator reduces some complexity.

Note that `mgr.cage()` requires `!glad` which implies that `tell` has already been called.

Comment (gbalabasquer) [reply src/mgr.sol:255]: Ah yes, you are right, so `cage` can't be called if governance hasn't previously called `tell`.
Tbh I need to draw a states flow chart to understand well all these cases.

Comment (MrChico) [reply src/mgr.sol:255]: Yeah, we should probably add one to the README

File: src/mgr.sol [hunk 10]
```diff
@@ -95,18 +95,18 @@ contract TinlakeManager {
     // Events
     event Rely(address usr);
     event Deny(address usr);
```
Comment (gbalabasquer) [inline src/mgr.sol]: Conversation got resolved but the `indexed` is still missing for `Rely` and `Deny`

Comment (xmxanuel) [reply src/mgr.sol]: fixed

File: src/mgr.sol [hunk 11]
```diff
@@ -110,33 +133,37 @@ contract TinlakeManager is LibNote {
...
-        daiJoin = GemJoinLike(daiJoin_);
+        end = EndLike(end_);
+        daiJoin = JoinLike(daiJoin_);
+        vow = vow_;
 
         gem = GemLike(drop_);
         pool = RedeemLike(pool_);
 
         ilk = ilk_;
         wards[msg.sender] = 1;
```
Comment (hexonaut) [inline src/mgr.sol:156]: Needs `emit Rely(msg.sender);` after this.

File: src/mgr.sol [hunk 12]
```diff
@@ -110,33 +133,37 @@ contract TinlakeManager is LibNote {
...
 
         vat = VatLike(vat_);
         dai = GemLike(dai_);
-        vow = VowLike(vow_);
-        daiJoin = GemJoinLike(daiJoin_);
+        end = EndLike(end_);
+        daiJoin = JoinLike(daiJoin_);
+        vow = vow_;
 
         gem = GemLike(drop_);
```
Comment (hexonaut) [inline src/mgr.sol:151]: Should add `require(gem.decimals() == dec)` for defensive programming.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Just some minor adjustments.

**Comment by hexonaut** [comment]
In general I would like to see some additional tests against the actual dss code instead of just mocks. Specifically the liquidation and write-off processes.

**Comment by xmxanuel** [comment]
@hexonaut thanks for your review you can find system tests in this repo:
https://github.com/centrifuge/tinlake-maker-tests

We have additional tests in the Tinlake core repo (https://github.com/centrifuge/tinlake/tree/develop/src/test/system/lender/mkr) where Maker is mocked but we re-run the same tests in `tinlake-maker-tests` with actual Maker contracts and the `mgr`

You can also find additional liquidation and write-offs tests in the test repo

File: src/mgr.sol (hunk 11 above)
Comment (godsflaw) [inline src/mgr.sol:156]:
```suggestion
        wards[msg.sender] = 1;
        emit Rely(msg.sender);
```

File: src/mgr.sol (hunk 12 above)
Comment (godsflaw) [inline src/mgr.sol:151]:
```suggestion
        gem = GemLike(drop_);
        require(gem.decimals() == dec, "TinlakeMgr/decimals-dont-match");
```

File: src/mgr.sol [hunk 13]
```diff
@@ -17,7 +17,6 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 pragma solidity >=0.5.12;
```
Comment (godsflaw) [inline src/mgr.sol]:
```suggestion
pragma solidity 0.5.12;
```

**Review by godsflaw — APPROVED** [review]
LGTM on accepted changes

**Comment by lucasvo** [comment]
Closing this PR as we're changing strategy a bit.

**Pull request closed without merging**